frame/lockdrop/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+        Campaigns get(fn campaigns): map hasher(blake2_128_concat) CampaignIdentifier => Option<CampaignInfo<T>>;
+        Locks get(fn locks): double_map hasher(blake2_128_concat) T::AccountId, hasher(blake2_128_concat) CampaignIdentifier => Option<LockInfo<T>>;
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+        /// The given campaign trying to create has already existed.
+        CampaignAlreadyExists,
+        /// Campaign end block must be in the future.
+        CampaignEndInThePast,
+        /// Campaign lock block must be after campaign end block.
+        CampaignLockEndBeforeCampaignEnd,
+        /// Campaign does not exist.
+        CampaignNotExists,
+        /// Invalid lock end block.
+        InvalidLockEndBlock,
+	}
+}
+
+decl_event! {
+	pub enum Event<T> where AccountId = <T as frame_system::Config>::AccountId {
+        CampaignCreated(CampaignIdentifier),
+        CampaignExpired(CampaignIdentifier),
+        Locked(CampaignIdentifier, AccountId),
+        Unlocked(CampaignIdentifier, AccountId),
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin { 
+        type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+        #[weight = 0]
+        fn create_campaign(origin, identifier: CampaignIdentifier, end_block: T::BlockNumber, min_lock_end_block: T::BlockNumber) {
+            ensure_root(origin)?;
+
+            ensure!(!Campaigns::<T>::contains_key(&identifier), Error::<T>::CampaignAlreadyExists);
+            
+            let current_number = frame_system::Module::<T>::block_number();
+            ensure!(end_block > current_number, Error::<T>::CampaignEndInThePast);
+            ensure!(min_lock_end_block > end_block, Error::<T>::CampaignLockEndBeforeCampaignEnd);
+            
+            Campaigns::<T>::insert(identifier, CampaignInfo { end_block, min_lock_end_block });
+            Self::deposit_event(Event::<T>::CampaignCreated(identifier));
+        }
+
+        #[weight = 0]
+        fn remove_expired_campaign(origin, identifier: CampaignIdentifier) {
+            ensure_signed(origin)?;
+
+            let info = Campaigns::<T>::get(&identifier);
+            if let Some(info) = info {
+                let current_number = frame_system::Module::<T>::block_number();
+                if current_number > info.end_block {
+                    Campaigns::<T>::remove(identifier);
+                    Self::deposit_event(Event::<T>::CampaignExpired(identifier));
+                }
+            }
+        }
+
+        #[weight = 0]
+        fn lock(origin, amount: BalanceOf<T>, identifier: CampaignIdentifier, lock_end_block: T::BlockNumber) {
+            let account_id = ensure_signed(origin)?;
+            let campaign_info = Campaigns::<T>::get(&identifier).ok_or(Error::<T>::CampaignNotExists)?;
+            
+            ensure!(lock_end_block > campaign_info.min_lock_end_block, Error::<T>::InvalidLockEndBlock);
```

Comment:
should also check that the current block is before the end block

Reply:
Ooops!